connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -307,15 +307,25 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
     }
 
     private void processTargetStateChanges(Set<String> connectorTargetStateChanges) {
-        if (!connectorTargetStateChanges.isEmpty()) {
-            for (String connector : connectorTargetStateChanges) {
-                if (worker.connectorNames().contains(connector)) {
-                    TargetState targetState = configState.targetState(connector);
-                    worker.setTargetState(connector, targetState);
-                    if (targetState == TargetState.STARTED)
-                        reconfigureConnectorTasksWithRetry(connector);
-                }
+        for (String connector : connectorTargetStateChanges) {
+            TargetState targetState = configState.targetState(connector);
+            if (targetState == null) {
+                // The target state was deleted. This should follow only after deletion of the
+                // connector's configuration itself, so raise an error if otherwise.
+                if (configState.connectors().contains(connector))
+                    throw new ConnectException("Unexpected removal of target state for connector " + connector);
```

Comment:
I don't think its safe to throw in this case. You can correctly arrive at this point by creating a connector, deleting it, and then recreating it. I think in general we may not handle this case well right now (e.g. for sink connectors their offsets are not reset), but this will just cause an exception to be thrown. Since we don't write the target state on configs (null is implicitly STARTED), we'll get to this point and throw an exception. Even if we did write the target state immediately after writing the config, we couldn't be certain they'd be read atomically so I think we could still hit this case.